nifi-nar-bundles/nifi-extension-utils/nifi-hadoop-utils/src/main/java/org/apache/nifi/hadoop/SecurityUtil.java

```diff
@@ -141,4 +143,37 @@ public static boolean isSecurityEnabled(final Configuration config) {
         Validate.notNull(config);
         return KERBEROS.equalsIgnoreCase(config.get(HADOOP_SECURITY_AUTHENTICATION));
     }
+
+    public static <T> T callWithUgi(CallableThrowingException<UserGroupInformation, IOException> ugiProvider, CallableThrowingException<T, IOException> function) throws IOException {
```

Comment:
In my opinion passing in a `UserGroupProvider` instance instead of the `ugiProvider` would make no difference from the functionality point of view but it'd simplify the usage/improve the readability.

As for the second parameter I'm not fully sure why we need the `CallableThrowingException` class. For me simple `Callable` seems to be ok too, even though it doesn't specify its exception type, but I don't see the need for that.

Reply:
wow, the first point from this comment has been fixed even before I submitted the comment :) Thanks @tpalfy

Reply:
I haven't looked too closely, but I was also wondering about the need for `CallableThrowingException `, isn't it basically providing the same thing as `PrivilegedExceptionAction` ?

I would think the code inside the client services could just be:
 ```
getUgi().doAs((PrivilegedExceptionAction<Object>)() -> {
   ...
}
```

Reply:
Yeah, that would work. I overcomplicated it a little it seems. Updated.